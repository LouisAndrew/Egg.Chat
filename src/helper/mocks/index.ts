export { mockChatroom } from './chatroom';
export { mockMessage } from './message';
export { mockUser2, mockUser1 } from './user';
export { mockFirebase } from './firebase';
