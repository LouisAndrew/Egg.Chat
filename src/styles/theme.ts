import buttons from './buttons';
import text from './texts';
import variants from './variants';

const theme = {
    breakpoints: ['40em', '52em', '64em'],
    fontSizes: [12, 14, 16, 20, 24, 32, 48, 64],
    colors: {
        blue: {
            //       third,    blue-shade, blue-darker, blue-even-darker
            dark: ['#545E75', '#4B556C', '#3A4765', '#2F3A54'],
            //      primary
            light: '#CEEDFF',
        },
        //       black-shade, text-color
        black: ['#A5A5A5', '#333333'],
        //       secondary,  bg
        white: ['#F3F3F3', '#F7F7FF'],
        //      banner-new, banner-online
        misc: ['#FF8590', '#ABFFB3'],
    },
    space: [0, 4, 8, 16, 32, 64, 128, 256],
    fonts: {
        body: "'Open Sans', sans-serif",
        heading: "'Raleway', sans-serif",
        monospace: 'Menlo, monospace',
    },
    fontWeights: {
        body: 400,
        heading: 700,
        bold: 700,
    },
    lineHeights: {
        body: 1.5,
        heading: 1.25,
    },
    shadows: {
        small: '0 0 4px rgba(0, 0, 0, .125)',
        large: '0 0 24px rgba(0, 0, 0, .125)',
    },
    variants,
    text,
    buttons,
};

export default theme;
