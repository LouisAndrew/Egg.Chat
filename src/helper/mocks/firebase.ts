// use just as reference.
// Error message: Jest mock must be an inline function

import { Chatroom, User } from '../schema';
import { mockUser1, mockUser2, mockChatroom } from '.';

const mockUserDb: { [key: string]: User } = {
    '1234': mockUser1,
    '5678': mockUser2,
};

const mockChatroomDb: { [key: string]: Chatroom } = {
    '0972': mockChatroom,
};

export const mockFirebase = () => ({
    firestore: () => ({
        collection: (ref: string) => {
            if (ref === 'user') {
                return {
                    doc: (id: string) => ({
                        get: () =>
                            new Promise((res) => {
                                res(mockUserDb[id]);
                            }),
                    }),
                };
            } else if (ref === 'chatroom') {
                return {
                    doc: (id: string) => ({
                        get: () =>
                            new Promise((res) => {
                                res(mockChatroomDb[id]);
                            }),
                    }),
                };
            } else {
                return;
            }
        },
    }),
    auth: () => ({}),
});
