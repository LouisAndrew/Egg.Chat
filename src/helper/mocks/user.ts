import { Status, User } from '../schema';

const CHATROOM_ID = '0972';

const mockUser1: User = {
    uid: '1234',
    displayImage: '',
    displayName: 'John Doe',
    status: Status.ONLINE,
    chatrooms: [CHATROOM_ID],
    lastOnline: new Date(16023160238),
};

const mockUser2: User = {
    uid: '5678',
    displayImage: '',
    displayName: 'Jane Doe',
    status: Status.OFFLINE,
    chatrooms: [CHATROOM_ID],
    lastOnline: new Date(1605460238),
};

export { mockUser2, mockUser1 };
