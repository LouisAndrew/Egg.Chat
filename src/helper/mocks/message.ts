import { Message } from '../schema';
import { mockUser1 } from './user';

const mockMessage: Message = {
    msgId: '45725',
    msg: 'Hello, World',
    sentBy: mockUser1.uid,
    sentAt: new Date(Date.parse('10 Oct 2020 00:12:00 GMT')),
};

export { mockMessage };
