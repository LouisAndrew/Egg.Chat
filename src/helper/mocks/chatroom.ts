import { Chatroom } from 'helper/schema';
import { mockUser2, mockUser1 } from './user';
import { mockMessage } from './message';

const CHATROOM_ID = '0972';

const mockChatroom: Chatroom = {
    roomId: CHATROOM_ID,
    users: [mockUser1.uid, mockUser2.uid],
    messages: [mockMessage],
};

export { mockChatroom, CHATROOM_ID };
