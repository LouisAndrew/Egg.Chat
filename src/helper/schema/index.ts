import { Chatroom } from './chatroom';
import { User, Status } from './user';
import { Message } from './message';

export { Status };
export type { Chatroom, User, Message };
