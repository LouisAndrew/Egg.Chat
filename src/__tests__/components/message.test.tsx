import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

import Message from 'components/message';

// mockMesasge is sent by user1.
import { mockMessage, mockUser1, mockUser2 } from 'helper/mocks';
import theme from 'styles/theme';
import WithTheme from 'helper/util/with-theme';

describe('Message', () => {
    const mockDeleteMessage = jest.fn((str) => {});

    // message: SENT
    const Element = (
        <WithTheme>
            <Message
                {...mockMessage}
                userId={mockUser1.uid}
                deleteMsg={mockDeleteMessage}
            />
        </WithTheme>
    );

    // message: RECEIVED.
    const ElementUser2 = (
        <WithTheme>
            <Message
                {...mockMessage}
                userId={mockUser2.uid}
                deleteMsg={mockDeleteMessage}
            />
        </WithTheme>
    );

    afterEach(cleanup);

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(Element, div);
    });

    it('Should render the message content correctly.', () => {
        const { queryByText } = render(Element);

        const content = queryByText(mockMessage.msg);
        expect(content).toBeInTheDocument();
    });

    it('Should render the "RECEIVED" variant of the component if the message is sent by logged in user.', () => {
        const { getByTestId } = render(ElementUser2);

        const wrapper = getByTestId('wrapper');
        expect(wrapper).toHaveStyle(
            `background-color: ${theme.colors.white[0]}`
        );
        expect(wrapper).toHaveStyle(`color: ${theme.colors.black[1]}`);
    });

    it('Should render the "SENT" variant of the component if message is not sent by logged in user', () => {
        const { getByTestId } = render(Element);

        const wrapper = getByTestId('wrapper');
        expect(wrapper).toHaveStyle(
            `background-color: ${theme.colors.blue.dark[0]}`
        );
        expect(wrapper).toHaveStyle(`color: ${theme.colors.white[1]}`);
    });

    it('Should not display menu-arrow when message is not sent by logged in user', () => {
        const { queryByTestId, getByTestId } = render(ElementUser2);

        const wrapper = getByTestId('wrapper');
        userEvent.hover(wrapper);

        const menuArrow = queryByTestId('menu-arrow');
        expect(menuArrow).not.toBeInTheDocument();
    });

    it('Should display menu-arrow when the component is hovered if the message is sent by logged in user.', () => {
        const { queryByTestId, getByTestId } = render(Element);

        const wrapper = getByTestId('wrapper');
        userEvent.hover(wrapper);

        const menuArrow = queryByTestId('menu-arrow');
        expect(menuArrow).toBeInTheDocument();
    });

    it('Should display menu to delete message when the menu-arrow is clicked.', () => {
        const { getByTestId } = render(Element);

        const wrapper = getByTestId('wrapper');
        userEvent.hover(wrapper);

        const menuArrow = getByTestId('menu-arrow');
        userEvent.click(menuArrow);

        const menuDelete = getByTestId('menu-delete');
        expect(menuDelete).toBeInTheDocument();
    });

    it('Should calls the deleteMsg when the menu to delete message is clicked.', () => {
        const { getByTestId } = render(Element);

        const wrapper = getByTestId('wrapper');
        userEvent.hover(wrapper);

        const menuArrow = getByTestId('menu-arrow');
        userEvent.click(menuArrow);

        const menuDelete = getByTestId('menu-delete');
        userEvent.click(menuDelete);

        expect(mockDeleteMessage).toBeCalled();
        expect(mockDeleteMessage).toBeCalledWith(mockMessage.msgId);
    });

    it('matches snapshot', () => {
        const run = true;

        if (run) {
            const tree = renderer.create(Element).toJSON();
            expect(tree).toMatchSnapshot();
        }
    });
});
