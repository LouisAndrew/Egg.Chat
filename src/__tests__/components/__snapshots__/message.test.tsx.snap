// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message matches snapshot 1`] = `
<div
  className="css-1kw0tju"
>
  <div
    className="css-pqyr7c"
  >
    02:12
  </div>
  <div
    className="css-8ma5b7"
    data-testid="wrapper"
  >
    <div
      className="css-vurnku"
    >
      Hello, World
    </div>
    <div
      className="menu-arrow css-1voq0am"
      data-testid="menu-arrow"
      onClick={[Function]}
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clipRule="evenodd"
          d="M11.354 1.646a.5.5 0 010 .708L5.707 8l5.647 5.646a.5.5 0 01-.708.708l-6-6a.5.5 0 010-.708l6-6a.5.5 0 01.708 0z"
          fillRule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;
